Lib/subprocess.py

```diff
@@ -968,12 +972,40 @@ def _execute_child(self, args, executable, preexec_fn, close_fds,
             # Process startup details
             if startupinfo is None:
                 startupinfo = STARTUPINFO()
-            if -1 not in (p2cread, c2pwrite, errwrite):
+
+            use_std_handles = -1 not in (p2cread, c2pwrite, errwrite)
+            if use_std_handles:
                 startupinfo.dwFlags |= _winapi.STARTF_USESTDHANDLES
                 startupinfo.hStdInput = p2cread
                 startupinfo.hStdOutput = c2pwrite
                 startupinfo.hStdError = errwrite
 
+            attribute_list = startupinfo.lpAttributeList
+            have_handle_list = (attribute_list and
+                                "handle_list" in attribute_list and
+                                attribute_list["handle_list"])
+
+            # If we were given an handle_list or need to create one
+            if have_handle_list or (use_std_handles and close_fds):
+                if attribute_list is None:
+                    attribute_list = startupinfo.lpAttributeList = {}
+                handle_list = attribute_list.setdefault("handle_list", [])
+
+                if use_std_handles:
+                    handle_list += [p2cread, c2pwrite, errwrite]
```

Comment:
Yes, I take for granted that we all know it's a bad thing to call `os.close` on a Windows handle. I noticed it because `self._closed_child_pipe_fds = True` isn't being set, but it needs to be fixed in both places because in principle instantiating `Popen` could fail before the try/finally in `_execute_child`.